README
WWW::Orphea
===========

NAME

    WWW::Orphea

DESCRIPTION

   This module may be used search images on an Orphea web server, a Digital
   Asset Management (DAM) software suite published by Algoba Systems
   (http://www.algoba.com/public/orphea.pgi). Its interface is adapted from
   WWW::Google::Images, itself inspired from WWW::Google::Groups.

HOW DO I INSTALL IT?

    To install this module type the following:

    tar xvzf WWW::Orphea-X.XX.tar.gz
    cd WWW::Orphea-X.XX
    perl Makefile.PL
    make
    make test
    make install

    Where X.XX is the version number of the module which you are installing.

WHERE IS THE DOCUMENTATION?

    All of the documentation is currently in POD format in the Images.pm
    file. If you install the module using the standard method you should  be
    able to read it by typing

    perldoc WWW::Orphea

    at a comand prompt.

LATEST VERSION

    The latest version of this module will always be available from CPAN.

COPYRIGHT AND LICENSE

    Copyright (C) 2004-2006, INRIA.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

AUTHOR

    Guillaume Rousse <grousse@cpan.org>


ChangeLog
2006-08-22 0.3.3

	* fixes for new Orphea version
	* make ExtUtils::MakeMaker::Coverage use conditional

2005-05-23 0.3.2
	
	* better test suite

2004-11-03 0.3.1
	
	* fixed documentation for saving method

2004-04-08 0.3

	* added Test::URI dependency
	* fixed ChangeLog

2004-04-05 0.2

	* skip tests if web unavailable

2004-04-02 0.1

	* first release


lib/WWW/Orphea/Image.pm
# $Id: Image.pm,v 1.5 2006/08/22 13:04:01 rousse Exp $
package WWW::Orphea::Image;

=head1 NAME

WWW::Orphea::Image - Image object for WWW::Orphea

=cut

=head1 Constructor

=head2 new(I<$agent>, I<$content>, I<$legend>)

Creates and returns a new C<WWW::Orphea::Image> object.

=cut

sub new {
    my ($class, $agent, $content, $legend) = @_;

    my $self = bless {
        _agent   => $agent,
        _content => $content,
        _legend  => $legend,
    }, $class;

    return $self;
}

=head1 Accessors

=head2 $image->content_url()

Returns the url to the image file itself.

=cut

sub content_url {
    my ($self) = @_;
    return $self->{_content};
}

=head2 $image->legend()

Returns the legend of the image.

=cut

sub legend {
    my ($self) = @_;
    return $self->{_legend};
}

=head1 Other methods

=head2 $image->save_content(I<%args>)

Save the image file. The default is to keep its original file name, but this behavior can be altered using optional parameters.

Optional parameters:

=over

=item file => I<$file>

Use $file as file name.

=item dir => I<$directory>

Use $directory as a directory path to save the file.  If only 'file' is specified, it will be saved in a path relative to the current working directory.

=item base => I<$base>

Use $base with .jpg extension added as file name.

=back

=cut

sub save_content {
    my ($self, %args) = @_;

    my $file = $self->_get_file($self->{_content}, %args);
    $self->_save_as($file, $self->{_content});

    return $file;
}

sub _save_as {
    my ($self, $file, $url) = @_;
    $self->{_agent}->get($url, ":content_file" => $file );
    $self->{_agent}->back();
}

sub _get_file {
    my ($self, $url, %args) = @_;

    my $file;
    if ($args{file}) {
        $file = $args{file};
    } elsif ($args{base}) {
        $file = $args{base} . '.jpg';
    } else {
        $url =~ /([^\/]+)$/;
        $file = $1;
    }

    my $dir;
    if ($args{dir}) {
        $dir = $args{dir};
    } else {
        $dir = '.';
    }

    return $dir . '/' . $file;
}

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2004-2006, INRIA.

This program is free software; you can redistribute it and/or modify it under the same terms as Perl itself.

=head1 AUTHOR

Guillaume Rousse <grousse@cpan.org>

=cut

1;


lib/WWW/Orphea/SearchResult.pm
# $Id: SearchResult.pm,v 1.4 2006/08/22 13:04:01 rousse Exp $
package WWW::Orphea::SearchResult;

=head1 NAME

WWW::Google::Orphea::SearchResult - Search result object for WWW::Orphea

=cut

use WWW::Orphea::Image;
use strict;

=head1 Constructor

=head2 new(I<$agent>, I<@urls>)

Creates and returns a new C<WWW::Google::Orphea::SearchResult> object.

=cut

sub new {
    my ($class, $agent, @images) = @_;

    my $self = bless {
        _agent  => $agent,
        _images => \@images
    }, $class;

    return $self;
}

=head1 Accessor

=head2 $result->next()

Returns the next image from result as a C<WWW::Orphea::Image> object.

=cut

sub next {
    my ($self) = @_;
    my $image = shift @{$self->{_images}};
    return unless $image;
    return WWW::Orphea::Image->new(
        $self->{_agent},
        $image->{content},
        $image->{legend}
    );
}

=head1 Other methods

=head2 $result->save_all_contents(I<%args>)

Save all the image files from result, by calling $image->save_content(I<%args>)
for each one. If optional parameter file or base is given, an index number is
automatically appended.

=cut

sub save_all_contents {
    my ($self, %args) = @_;

    my $count;
    while (my $image = $self->next()) {
        $count++;
        $image->save_content(
            dir  => $args{dir} ? $args{dir} : undef,
            file => $args{file} ? $args{file} . $count : undef,
            base => $args{base} ? $args{base} . $count : undef,
        );
    }
}

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2004-2006, INRIA.

This program is free software; you can redistribute it and/or modify it under the same terms as Perl itself.

=head1 AUTHOR

Guillaume Rousse <grousse@cpan.org>

=cut

1;


lib/WWW/Orphea.pm
# $Id: Orphea.pm,v 1.15 2006/08/22 13:00:51 rousse Exp $
package WWW::Orphea;

=head1 NAME

WWW::Orphea - Orphea Agent

=head1 VERSION

Version 0.3.3

=head1 DESCRIPTION

This module may be used search images on an Orphea web server, a Digital Asset
Management (DAM) software suite published by Algoba Systems
(http://www.algoba.com/public/orphea.pgi). Its interface is adapted from
L<WWW::Google::Images>, itself inspired from L<WWW::Google::Groups>.

=head1 SYNOPSIS

    use WWW::Orphea;

    $agent = WWW::Orphea->new(
        server => 'my.orphea.server',
        proxy  => 'my.proxy.server:port',
    );

    $result = $agent->search('flowers', limit => 10);

    while ($image = $result->next()) {
        $count++;
        print $image->content_url();
        print $image->legend();
        print $image->save_content( base => 'image' . $count);
    }

=cut

use WWW::Mechanize;
use WWW::Orphea::SearchResult;
use strict;
our $VERSION = '0.3.3';

=head1 Constructor

=head2 new(I<%args>)

Creates and returns a new C<WWW::Google::Images> object.

Optional parameters:

=over

=item server => I<$server>

use I<$server> as server.

=item proxy => I<$proxy>:I<$port>

use I<$proxy> as proxy on port I<$port>.

=back

=cut

sub new {
    my ($class, %arg) = @_;

    foreach my $key (qw(server pass user)) {
        die "No $key defined, aborting" unless $arg{$key};
    }

    foreach my $key (qw(server proxy)) {
        next unless $arg{$key};
        $arg{$key} = 'http://' . $arg{$key} unless $arg{$key} =~ m|^\w+?://|;
        $arg{$key} = $arg{$key} . '/' unless $arg{$key} =~ m|/$|;
    }

    my $a = WWW::Mechanize->new(onwarn => undef, onerror => undef);
    $a->proxy(['http'], $arg{proxy}) if $arg{proxy};

    my $self = bless {
        _user   => $arg{user},
        _pass   => $arg{pass},
        _server => $arg{server},
        _proxy  => $arg{proxy},
        _agent  => $a,
    }, $class;

    $self->{_agent}->get($self->{_server});

    $self->{_agent}->submit_form(
        form_number => 1,
        fields      => {
            UserName => $self->{_user},
            PassWord => $self->{_pass},
        }
    );

    $self->{_agent}->content() =~ /src="loading_header.html\?UNID=(\w+)&LGID=(\w+)&([^"]+)"/;
    my $unid   = $1;
    my $lang   = $2;
    my $remain = $3;
    $self->{_agent}->get($self->{_server} . "header.html?UNID=$unid&LGID=$lang&$remain");

    my $content = $self->{_agent}->content();
    $content =~ s/document.write\('//g;
    $content =~	s/'\);//g;
    $content =~	s/\\'/'/g;
    $content =~	s/'\+lvLangue\+'/$lang/g;
    $content =~ s/'\+lvUNID\+'/$unid/g;
    $content =~ s/<!-- ORPHEA CODE INSERTED  -->/<\/script>/;
    $self->{_form} = HTML::Form->parse($content, $self->{_agent}->uri());

    return $self;
}

=head2 $agent->search(I<$query>, I<%args>);

Perform a search for I<$query>, and return a C<WWW::Google::Images::SearchResult> object.

Optional parameters:

=over

=item limit => I<$limit>

limit the maximum number of result returned to $limit.

=back

=cut

sub search {
    my ($self, $query, %arg) = @_;

    warn "No query given, aborting" and return unless $query;

    $arg{limit} = 10 unless defined $arg{limit};

    $self->{_form}->value('userrequest', $query);
    $self->{_agent}->request($self->{_form}->click());
    $self->{_agent}->submit();

    my @images;
    my $page = 1;

    LOOP: {
        do {
            push(@images, $self->_extract_images($arg{limit} ? $arg{limit} - @images : 0));
            last if $arg{limit} && @images >= $arg{limit};
        } while ($self->_next_page(++$page));
    }

    return WWW::Orphea::SearchResult->new($self->{_agent}, @images);
}

sub _next_page {
    my ($self, $page) = @_;

    my $link = $self->{_agent}->find_link(url_regex => qr/javascript:fOpenURL/, text_regex => qr/$page/);
    return unless $link;
    my $url = $link->url();
    $url =~ s/javascript:fOpenURL\('//;
    $url =~ s/'\)//;
    return $self->{_agent}->get($self->{_server} . $url);
}

sub _extract_images {
    my ($self, $limit) = @_;

    my @images;
    my $page = $self->{_agent}->content();
    my @legends  = $page =~ m/<TD class="photoslistsubline1">([^<]+)<\/TD>/go;
    my @contents = $page =~ m/<img src="thu_orphea\/([\d_]+)\.(?:thw|THW)"/go;

    for (my $i = 0; $i <= $#contents; $i++) {
        last if $limit && @images >= $limit;
        $contents[$i] = $self->{_server} . 'bro_orphea/' . $contents[$i] . '.BRO';
        $legends[$i]  =~ s/\w+$//;
        push(@images, { content => $contents[$i], legend => $legends[$i] });
    }

    return @images;
}

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2004-2006 INRIA.

This program is free software; you can redistribute it and/or modify it under the same terms as Perl itself.

=head1 AUTHOR

Guillaume Rousse <grousse@cpan.org>

=cut

1;
